### PR TITLE
bots: Implement cluster based machine learning

### PR DESCRIPTION
**Pull request by stefwalter**
This attempts to cluster log items related to similarity and then
classify whether new test logs fit into those clusters. The clustering
is unsupervised, and currently uses DBSCAN to accomplish this.
The classification currently uses nearest neighbor techniques.
    
We use distances to tell us whether two items are similar or not.
These distances are currently calculated via normalized compression
distance in ncd.py
    
We run this code in addition to the older Neural Network machine
learning so we can compare the difference in effectiveness.
    
The machine learning uploads the clusters it finds to the logs
when it performs its job.
    
The tests-policy script has been updated to print out the likelyhood
that a test failure is a flake.

 * [x] learn-tests

### CODE REVIEW

**Comment by stefwalter** [comment]
Running once locally, and uploading my results. Then will remove WIP

**Comment by cockpituous** [comment]
learn-tests in progress on cockpit-tasks-97kg9.
Log: http://fedorapeople.org/groups/cockpit/logs/learn-tests-8380-20180109-200202/

**Comment by cockpituous** [comment]
learn-tests in progress on cockpit-tasks-7f5gf.
Log: http://fedorapeople.org/groups/cockpit/logs/learn-tests-8380-20180109-200303/

**Comment by cockpituous** [comment]
tests-data command failed: http://fedorapeople.org/groups/cockpit/logs/learn-tests-8380-20180109-200303/

**Comment by cockpituous** [comment]
learn-tests in progress on cockpit-tasks-73q6p.
Log: http://fedorapeople.org/groups/cockpit/logs/learn-tests-8380-20180110-064343/

**Comment by cockpituous** [comment]
learn-tests in progress on cockpit-tasks-h0jkf.
Log: http://fedorapeople.org/groups/cockpit/logs/learn-tests-8380-20180110-092323/

**Comment by cockpituous** [comment]
tests-data command failed: http://fedorapeople.org/groups/cockpit/logs/learn-tests-8380-20180110-092323/

**Comment by cockpituous** [comment]
learn-tests in progress on cockpit-tasks-9tl3b.
Log: http://fedorapeople.org/groups/cockpit/logs/learn-tests-8380-20180110-100404/

**Comment by cockpituous** [comment]
tests-data command failed: http://fedorapeople.org/groups/cockpit/logs/learn-tests-8380-20180110-100404/

**Comment by cockpituous** [comment]
learn-tests in progress on verifymachine4.
Log: http://fedorapeople.org/groups/cockpit/logs/learn-tests-8380-20180110-101717/

**Comment by cockpituous** [comment]
learn-tests in progress on cockpit-tasks-b6lvv.
Log: http://fedorapeople.org/groups/cockpit/logs/learn-tests-8380-20180110-124646/

**Comment by cockpituous** [comment]
learn-tests in progress on cockpit-tasks-9tl3b.
Log: http://fedorapeople.org/groups/cockpit/logs/learn-tests-8380-20180110-142424/

**Comment by cockpituous** [comment]
learn-tests in progress on cockpit-tasks-n9dsb.
Log: http://fedorapeople.org/groups/cockpit/logs/learn-tests-8380-20180110-161010/

**Comment by cockpituous** [comment]
learn-tests in progress on cockpit-tasks-hd9gc.
Log: http://fedorapeople.org/groups/cockpit/logs/learn-tests-8380-20180110-173232/

**Comment by cockpituous** [comment]
learn-tests in progress on cockpit-tasks-hd9gc.
Log: http://fedorapeople.org/groups/cockpit/logs/learn-tests-8380-20180110-185656/

**Comment by cockpituous** [comment]
learn-tests in progress on cockpit-tasks-tztsr.
Log: http://fedorapeople.org/groups/cockpit/logs/learn-tests-8380-20180110-205252/

**Comment by cockpituous** [comment]
learn-tests in progress on cockpit-tasks-n8pm7.
Log: http://fedorapeople.org/groups/cockpit/logs/learn-tests-8380-20180110-221818/

**Comment by cockpituous** [comment]
learn-tests in progress on cockpit-tasks-n9dsb.
Log: http://fedorapeople.org/groups/cockpit/logs/learn-tests-8380-20180110-221818/

**Comment by cockpituous** [comment]
learn-tests in progress on cockpit-tasks-qddr1.
Log: http://fedorapeople.org/groups/cockpit/logs/learn-tests-8380-20180110-233333/

**Comment by cockpituous** [comment]
learn-tests in progress on cockpit-tasks-0rt9w.
Log: http://fedorapeople.org/groups/cockpit/logs/learn-tests-8380-20180111-010202/

**Comment by cockpituous** [comment]
learn-tests in progress on verifymachine4.
Log: http://fedorapeople.org/groups/cockpit/logs/learn-tests-8380-20180111-015858/

**Comment by cockpituous** [comment]
learn-tests in progress on cockpit-tasks-73q6p.
Log: http://fedorapeople.org/groups/cockpit/logs/learn-tests-8380-20180111-035858/

**Comment by cockpituous** [comment]
learn-tests in progress on cockpit-tasks-0rt9w.
Log: http://fedorapeople.org/groups/cockpit/logs/learn-tests-8380-20180111-052020/

**Comment by cockpituous** [comment]
learn-tests in progress on cockpit-tasks-n8pm7.
Log: http://fedorapeople.org/groups/cockpit/logs/learn-tests-8380-20180111-061010/

**Comment by cockpituous** [comment]
learn-tests in progress on verifymachine4.
Log: http://fedorapeople.org/groups/cockpit/logs/learn-tests-8380-20180111-071717/

**Comment by cockpituous** [comment]
learn-tests in progress on cockpit-tasks-9wwzc.
Log: http://fedorapeople.org/groups/cockpit/logs/learn-tests-8380-20180111-081818/

**Comment by cockpituous** [comment]
learn-tests in progress on cockpit-tasks-b6lvv.
Log: http://fedorapeople.org/groups/cockpit/logs/learn-tests-8380-20180111-090606/

File: bots/task/sink.py [hunk 1]
```diff
@@ -53,7 +54,7 @@ def __init__(self, host, identifier, status=None):
         os.dup2(self.ssh.stdin.fileno(), 2)
 
     def attach(self, filename):
-        shutil.move(filename, self.attachments)
+        shutil.copy(filename, self.attachments)
```
[the remark this answers is not in the document]
Comment (stefwalter) [reply bots/task/sink.py]: This is actually unused. Should I remove it?

File: bots/image-download [hunk 2]
```diff
@@ -92,33 +97,46 @@ def download(dest, force, quiet, stores):
...
+                    sys.stderr.write(" > {0}\n".format(message))
+                continue
             except subprocess.CalledProcessError:
                 pass
 
-
-        # If the above block found nothing, then continue with next store
-        else:
             if not quiet:
                 sys.stderr.write(" x {0}\n".format(message))
```
[the remark this answers is not in the document]
Comment (stefwalter) [reply bots/image-download]: I don't think it's redundant. The 'x' shows that it wasn't found at the given IP adddress.

File: bots/task/sink.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply bots/task/sink.py]: Ah, so much the better, so nothing else uses this "old" method any more then?  Then it should get removed indeed.

File: bots/image-download (hunk 2 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply bots/image-download]: Oh, I see, for the case where the above does not do a `continue`; ok then.

**Comment by martinpitt** [comment]
rhel-atomic passed, but it wasn't noticed here due to the github outage . fedora-atomic failure is unrelated.

**Pull request merged**